### PR TITLE
Adding handling of TextInputClient.onConnectionClosed messages handli…

### PR DESCRIPTION
**Pull request by nturgut**
Adding handling of TextInputClient.onConnectionClosed messages handling to Framework

## Description

Adding a message to the Flutter Framework for clients to call if a connection is closed.

Information in following situations can be synced with the framework using this method call.

- if the mobile browser is sent to background by pressing the home button on a mobile device.
- If the blur event is triggered on an input element in a web browser.
- the user closed the virtual keyboard indication they stopped editing.

This PR is merged and rolled back before: https://github.com/flutter/flutter/pull/35100

## Related Issues

https://github.com/flutter/flutter/issues/43034

## Tests

I added the following tests:

- Unit tests for editable_text_test.dart
- Unit tests for text_input.dart

Manually tested with the Flutter Web Engine.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

Announcement: https://groups.google.com/forum/#!topic/flutter-announce/IwG8xGA7Kmo

### CODE REVIEW

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (mdebbar) [inline packages/flutter/test/widgets/editable_text_test.dart]: Another interesting test would be to use the text field again after the connection is closed (it should create a new connection and work fine).

**Review by mdebbar — COMMENTED** [review]
Thanks for tackling this!

Could you please double-check that this PR doesn't introduce the same regressions that caused the other PR to be reverted? cc @matthew-carroll

**Comment by matthew-carroll** [comment]
I think we may have said we were waiting for new test infra to attempt another solution to this due to the nuanced ways a fix might introduce new bugs, but I'll defer to @HansMuller @gspencergoog and @collinjackson on that topic.

File: packages/flutter/lib/src/services/text_input.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/text_input.dart]: Is this a request to close the connection (and the framework responds with actually closing the connection)? Or is this to inform the framework that the connection has been closed already (and the framework should not respond with closing it again)?

File: packages/flutter/test/services/text_input_test.dart
Comment (goderbauer) [inline packages/flutter/test/services/text_input_test.dart]: nit: formatting here seems off?

File: packages/flutter/lib/src/services/text_input.dart
Comment (goderbauer) [reply packages/flutter/lib/src/services/text_input.dart]: Right now its name suggests the latter, but the responds is currently the former, I think?

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (goderbauer) [reply packages/flutter/test/widgets/editable_text_test.dart]: +1

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (goderbauer) [inline packages/flutter/test/widgets/editable_text_test.dart]: Why is this method async and returning a Future?

Comment (goderbauer) [reply packages/flutter/test/widgets/editable_text_test.dart]: Also, what do we benefit from actually defining this method? The test seems clearer to my by just putting the expect directly where this method is called?

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (goderbauer) [inline packages/flutter/test/widgets/editable_text_test.dart]: Why this?

File: packages/flutter_test/lib/src/test_text_input.dart
Comment (goderbauer) [inline packages/flutter_test/lib/src/test_text_input.dart]: nit: this is not a full sentence.

**Comment by goderbauer** [comment]
@matthew-carroll From the comments on https://github.com/flutter/flutter/pull/39523 it sounded like the issue that caused the rollback was how we send this event from Android?

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: Sure, I'll delete it.

There was another one like this in the same file, so I thought it is common practice to define a verify method first.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: I want to it to be as close to the previous test as possible. I usually find it easier to troubleshoot when there is an error, if some other tests also have the same setup.

I can change it to another widget if you want to :) 

One thing I'll add is also: "switching between two editable text fields after connection closed". I believe it might be good to test that case.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: Sure, it sounds very useful I'll add it.

File: packages/flutter/lib/src/services/text_input.dart
Comment (nturgut) [reply packages/flutter/lib/src/services/text_input.dart]: Shall we use a method name like "ConnectionCloRequest"? 

@mdebbar do you have a suggestion?

Comment (mdebbar) [reply packages/flutter/lib/src/services/text_input.dart]: I personally prefer the latter (the platform message informs the framework that the connection has been closed, and the framework doesn't respond).

For the web, both approaches are possible, I believe. Not sure about Android and iOS.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: Done.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: done.

File: packages/flutter/lib/src/services/text_input.dart
Comment (nturgut) [reply packages/flutter/lib/src/services/text_input.dart]: Thanks I'll continue with the latter(connectionClosed), then.

File: packages/flutter_test/lib/src/test_text_input.dart
Comment (nturgut) [reply packages/flutter_test/lib/src/test_text_input.dart]: done.

File: packages/flutter/lib/src/services/text_input.dart
Comment (goderbauer) [reply packages/flutter/lib/src/services/text_input.dart]: The current implementation calls _closeInputConnectionIfNeeded, though. Which will make the frame respond by closing the connection?

Also: We need to double-check that this model would work on other platforms like iOS and Android.

File: packages/flutter/test/services/text_input_test.dart
Comment (nturgut) [reply packages/flutter/test/services/text_input_test.dart]: done.

File: packages/flutter/lib/src/services/text_input.dart
Comment (nturgut) [reply packages/flutter/lib/src/services/text_input.dart]: The implementation of _closeInputConnectionIfNeeded:
void _closeInputConnectionIfNeeded() {
    if (_hasInputConnection) {
      _textInputConnection.close();
      _textInputConnection = null;
      _lastKnownRemoteTextEditingValue = null;
    }
  }

I believe this implementation is correct and this is what we need to do for closing the connection after the platform message is received. 

Shall I rename this method instead? Am I missing anything?

Comment (mdebbar) [reply packages/flutter/lib/src/services/text_input.dart]: `_textInputConnection.close();` sends a platform message (`TextInput.clearClient`) back to the engine asking it to close the connection. But since the engine is the one that requested the connection to be closed, there's no need for the framework to send the `TextInput.clearClient` message.

**Comment by nturgut** [comment]
We chat with @matthew-carroll  and @gspencergoog, and decided to continue with the framework change for now. Android change will follow when the integration tests for testing the behavior is supported.

**Comment by matthew-carroll** [comment]
@nturgut the update here based on @xster's comment is that if we prioritize this for the upcoming code freeze then we will probably have to implement the Android fix before we have new testing infra. I think web can continue as discussed, but we may be bringing in the Android side earlier than anticipated.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (mdebbar) [inline packages/flutter/lib/src/widgets/editable_text.dart]: I'm curious why not inline this method here?

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (mdebbar) [inline packages/flutter/test/widgets/editable_text_test.dart]: Does this `expect()` add any value?

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (mdebbar) [inline packages/flutter/test/widgets/editable_text_test.dart]: It might be useful to also check that the framework didn't respond with any platform messages.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (mdebbar) [inline packages/flutter/test/widgets/editable_text_test.dart]: nit: extra `//`

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (mdebbar) [inline packages/flutter/test/widgets/editable_text_test.dart]: Hmmm I wonder if there's a better way to check the focus state of a widget. @gspencergoog ?

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (mdebbar) [inline packages/flutter/test/widgets/editable_text_test.dart]: Are there any platform messages we could check here?

File: packages/flutter_test/lib/src/test_text_input.dart
Comment (mdebbar) [inline packages/flutter_test/lib/src/test_text_input.dart]: nit: extra trailing comma

**Review by mdebbar — APPROVED** [review]
LGTM with some comments about the tests.

File: packages/flutter/lib/src/services/text_input.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/text_input.dart]: Can we add a little more documentation here: e.g. when this notification is received the connection should no longer be used or something like that.

Also, what's meant by "due to user activity"? This can be sent whenever the platform wants to close the connection - even if that is not in response to user activity, no?

File: packages/flutter/lib/src/services/text_input.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/text_input.dart]: This doc is also confusing. This line says the user closed the connection, the next line says its called in responds to a platform notification.

Also, this method should document how this close is different from the close method above and when to call what.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/editable_text.dart]: It would probably be good to document the difference between this and the previous method for future readers.

**Review by goderbauer — COMMENTED** [review]
Approach looks good.

Added some comments about documentation. Since the text connection is a delicate dance between framework and engine to keep everything in sync we should document very clearly how things are intended to be used.

File: packages/flutter/lib/src/services/text_input.dart
Comment (dnfield) [reply packages/flutter/lib/src/services/text_input.dart]: How will we differentiate between the platform sending this because it's about to change the style of the keyboard, and the platform doing this because the user has asked it to dismiss the keyboard permanently?

File: packages/flutter/lib/src/services/text_input.dart
Comment (dnfield) [inline packages/flutter/lib/src/services/text_input.dart]: How does this not suffer from the same bugs that led to reverting https://github.com/flutter/flutter/pull/39523/files

**Review by dnfield — DISMISSED** [review]
This appears to re-introduce bugs for Android from the linked PR.

We should not land this without a test asserting that text input works as expected on Android with these changes.

File: packages/flutter/lib/src/services/text_input.dart
Comment (nturgut) [reply packages/flutter/lib/src/services/text_input.dart]: We don't think this will break android since:

(1) This is an optional channel, we will use it on web, android won't get effected as long as it is not implemented. 
(2) The implementations are not the same. In the previous implementation the connection was getting closed but the editabletext widget's state was not changed. In this implementation as well as closing the connection, these two lines are added in editabletext:
_lastKnownRemoteTextEditingValue = null;
 _finalizeEditing(true);

**Comment by nturgut** [comment]
> This appears to re-introduce bugs for Android from the linked PR.
> 
> We should not land this without a test asserting that text input works as expected on Android with these changes.

We already discussed it with @matthew-carroll @goderbauer @gspencergoog it is safe to land

**Review by dnfield — COMMENTED** [review]
After looking a bit more at this with @nturgut, this can't cause problems for Android (or iOS) becuase those embeddings don't ever call this method at this point.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: Yes it shows that the clear or hide messages weren't called.

When clearClient/hideClient are called, it is set to false. This is the code for the test_text_input.dart:

https://github.com/flutter/flutter/blob/6bf1ad4bd5714f73c1c808dc821add7199260d0c/packages/flutter_test/lib/src/test_text_input.dart#L86

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: The expect for visibility does that since the code of test_text_input is written that way.

https://github.com/flutter/flutter/blob/6bf1ad4bd5714f73c1c808dc821add7199260d0c/packages/flutter_test/lib/src/test_text_input.dart#L86

as long as test_text_input.isVisible true it means that clearClient or hide platform messages are not sent.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: I wasn't able to find one in the test doubles :)  It will be great if there is one!

I can also do write the validation as if it is more readbale
`expect(state.widget.focusNode.hasFocus, true)
`

File: packages/flutter/lib/src/services/text_input.dart
Comment (nturgut) [reply packages/flutter/lib/src/services/text_input.dart]: The platform messages sent by the engines (Android, Web etc) are called TextInputClient.methodName. Such as TextInputClient.performAction or TextInputClient.updateEditingState

The framework send methods are TextInput.methodName such as TextInput.clearClient, TextInput.hide, TextInput.show

Thanks for bringing this up!

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (nturgut) [reply packages/flutter/lib/src/widgets/editable_text.dart]: done. I also carried the method. I want all close/open related methods to be close, to group a subject area.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: done.

File: packages/flutter_test/lib/src/test_text_input.dart
Comment (nturgut) [reply packages/flutter_test/lib/src/test_text_input.dart]: done.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: I added the check.

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: I changed the expect to be more readable. I added a log to testTextInput

File: packages/flutter/test/widgets/editable_text_test.dart
Comment (nturgut) [reply packages/flutter/test/widgets/editable_text_test.dart]: done.

File: packages/flutter/lib/src/services/text_input.dart
Comment (nturgut) [reply packages/flutter/lib/src/services/text_input.dart]: Added more comments.

File: packages/flutter/lib/src/services/text_input.dart
Comment (nturgut) [reply packages/flutter/lib/src/services/text_input.dart]: done.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (nturgut) [reply packages/flutter/lib/src/widgets/editable_text.dart]: I removed this method. Only "connectionClosed" remains now which is documented on the interface.

File: packages/flutter/lib/src/services/text_input.dart
Comment (nturgut) [reply packages/flutter/lib/src/services/text_input.dart]: done. Also updated the comments.

File: packages/flutter/lib/src/services/text_input.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/text_input.dart]: Sorry, I forgot to notice this before, but this looks like a breaking change. Subclasses/Implementors of TextInputClient will have to be changed. We should follow the breaking change process, maybe since up with Dan since he also has a breaking change in this area: https://github.com/flutter/flutter/pull/43959

File: packages/flutter/lib/src/services/text_input.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/text_input.dart]: system_channels.dart documents the messages a channel uses. Can you please add that there?

**Review by goderbauer — COMMENTED** [review]
This looks fine (modulo the breaking change requirements).

However, there are two similar PRs out now. The other one is https://github.com/flutter/flutter/pull/43959. Just to double check: Do the addressed use cases really need two new APIs?

File: packages/flutter/lib/src/services/text_input.dart
Comment (nturgut) [reply packages/flutter/lib/src/services/text_input.dart]: Thanks for noticing this. I synced with @dnfield I'll send the email for both of us now.

File: packages/flutter/lib/src/services/text_input.dart
Comment (nturgut) [reply packages/flutter/lib/src/services/text_input.dart]: done. Thanks for the suggestion!

File: packages/flutter/lib/src/services/system_channels.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/system_channels.dart]: used? Oh, maybe "the user closed..."?

File: packages/flutter/lib/src/services/system_channels.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/system_channels.dart]: is not called -> must not be called?

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by goderbauer** [comment]
Please link the breaking change announcement from the PR description.

**Comment by nturgut** [comment]
> Please link the breaking change announcement from the PR description.

Done! Thanks.

**Pull request merged**
